RecoEgamma/EgammaTools/src/HGCalShowerShapeHelper.cc

```diff
@@ -0,0 +1,287 @@
+#include "RecoEgamma/EgammaTools/interface/HGCalShowerShapeHelper.h"
+
+const double HGCalShowerShapeHelper::kLDWaferCellSize_ = 0.698;
+const double HGCalShowerShapeHelper::kHDWaferCellSize_ = 0.465;
+
+void HGCalShowerShapeHelper::initPerEvent(const edm::EventSetup &iSetup, const std::vector<reco::PFRecHit> &pfRecHits) {
+  edm::ESHandle<CaloGeometry> geom;
+  iSetup.get<CaloGeometryRecord>().get(geom);
+  recHitTools_.setGeometry(*(geom.product()));
+
+  setPFRecHitPtrMap(pfRecHits);
+}
+
+void HGCalShowerShapeHelper::initPerObject(const std::vector<std::pair<DetId, float> > &hitsAndFracs,
+                                           double minHitE,
+                                           double minHitET,
+                                           int minLayer,
+                                           int maxLayer,
+                                           DetId::Detector subDet) {
+  maxLayer = maxLayer <= 0 ? recHitTools_.lastLayerEE() : maxLayer;
+
+  // Safety checks
+  nLayer_ = maxLayer - minLayer + 1;
+  assert(nLayer_ > 0);
+
+  minHitE_ = minHitE;
+  minHitET_ = minHitET;
+  minHitET2_ = minHitET * minHitET;
+  minLayer_ = minLayer;
+  maxLayer_ = maxLayer;
+  subDet_ = subDet;
+
+  setFilteredHitsAndFractions(hitsAndFracs);
+
+  setLayerWiseStuff();
+}
+
+void HGCalShowerShapeHelper::setPFRecHitPtrMap(const std::vector<reco::PFRecHit> &recHits) {
+  pfRecHitPtrMap_.clear();
+
+  for (const auto &recHit : recHits) {
+    pfRecHitPtrMap_[recHit.detId()] = &recHit;
+  }
+}
+
+void HGCalShowerShapeHelper::setFilteredHitsAndFractions(const std::vector<std::pair<DetId, float> > &hitsAndFracs) {
+  hitsAndFracs_.clear();
+  hitEnergies_.clear();
+  hitEnergiesWithFracs_.clear();
+
+  for (const auto &hnf : hitsAndFracs) {
+    DetId hitId = hnf.first;
+    float hitEfrac = hnf.second;
+
+    int hitLayer = recHitTools_.getLayer(hitId);
+
+    if (hitLayer > nLayer_) {
+      continue;
+    }
+
+    if (hitId.det() != subDet_) {
+      continue;
+    }
+
+    if (pfRecHitPtrMap_.find(hitId.rawId()) == pfRecHitPtrMap_.end()) {
+      continue;
+    }
+
+    reco::PFRecHit recHit = *pfRecHitPtrMap_[hitId.rawId()];
```

Comment:
```suggestion
    const reco::PFRecHit& recHit = *pfRecHitPtrMap_[hitId.rawId()];
```
to avoid making a copy of the rechit.